### PR TITLE
Implement a flag to enable inlining of formulas

### PR DESCRIPTION
**Pull request by facundominguez**
Follow up from #473 

Inlining ANF bindings can either speed up or slow down your verification. `Fusion.T.hs` is a benchmark in LH that improves and `FoldrUniversal.hs` is a benchmark that worsens.

### CODE REVIEW

**Comment by facundominguez** [comment]
`Fusion.T.hs` takes near 30 seconds without inlining, succeeds in 12 seconds with `--inline-anf-bindings=2`, and fails with `--inline-anf-bindings=3` as follows:
```
          > **** LIQUID: ERROR :1:1-1:1: Error
          >   elaborate solver elabBE 501 "lq_anf$##7205759403792808319##d3Pp" {lq_tmp$x##1854 : [Char] | [((len lq_tmp$x##1854) = ((len lit$36$Data.ByteString.Lazy$58$$32$invariant$32$violation$58$) + (len lq_anf$##7205759403792808318##d3Po)));
          >                             ((len lq_tmp$x##1854) >= 0)]} failed on:
          >       len lq_tmp$x##1854 == len "Data.ByteString.Lazy: invariant violation:" + len lq_anf$##7205759403792808318##d3Po
          >       && len lq_tmp$x##1854 >= 0
          >   with error
          >       Cannot unify (@(44) @(45)) with Str in expression: len "Data.ByteString.Lazy: invariant violation:" 
          >   in environment
          >       len := func(2 , [(@(0) @(1)); int])
          > 
          >       lq_anf$##7205759403792808318##d3Po := [Char]
          > 
          >       lq_tmp$x##1854 := [Char]
```

**Comment by facundominguez** [comment]
This is ready for review. I added unit tests to find a bug in the implementation of environment reduction that prevented LH from having green CI.

File: tests/tasty/ShareMapReference.hs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import           Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HashMap
+import           Data.HashSet (HashSet)
+import qualified Data.HashSet as HashSet
+import           Data.List (find)
+
+-- | See "Data.ShareMap" for documentation.
+data ShareMap k v = ShareMap
+  { toHashMap :: HashMap k v
+  , keyPartitions :: [HashSet k]
```
Comment (facundominguez) [inline tests/tasty/ShareMapReference.hs]: Document invariants
```suggestion
  { -- | Contains all the values in the ShareMap.
    toHashMap :: HashMap k v
  , -- | Contains all the keys in the ShareMap.
    --
    -- All sets in @keyPartitions@ are disjoint.
    -- Every key in in @keyPartitions@ is present in @toHashMap@.
    -- If @k1@ and @k2@ belong to the same set @s@ in @keyPartitions@
    -- then @toHashMap ! k1 == toHashMap ! k2@.
    keyPartitions :: [HashSet k]
```

**Review by ranjitjhala — APPROVED** [review]

**Comment by ranjitjhala** [comment]
Looks great to me, thank you @facundominguez !!! shall I merge it?

**Comment by facundominguez** [comment]
Yes! It is good to go.

**Pull request merged**